### PR TITLE
Add List box

### LINKED ISSUE #2876: List

### PR DESCRIPTION
**Pull request by joshwooding**
Closes #2876

### CODE REVIEW

File: packages/core/src/list-box/ListBox.tsx [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+
+export type ListBoxProps<Item = string> = {
+  /**
+   * If `true`, the list box will be disabled.
+   */
+  disabled?: boolean;
+  /**
+   * If `true`, the list box will have not a border.
+   */
+  borderless?: boolean;
```
Comment (origami-z) [inline packages/core/src/list-box/ListBox.tsx:32]: Small question, are we going to make this the only component do `borderless`, v.s. form controls doing `bordered`

File: packages/core/src/list-box/ListBox.tsx [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+    select,
+    setFocusVisibleState,
+    setFocusedState,
+    listRef,
+  } = listControl;
+
+  const typeaheadString = useRef("");
+  const typeaheadTimeout = useRef<number | undefined>();
+
+  const handleTypeahead = (event: KeyboardEvent<HTMLDivElement>) => {
```
Comment (origami-z) [inline packages/core/src/list-box/ListBox.tsx:99]: This may not be covered in test?

File: packages/core/src/list-box/ListBox.tsx [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+    // We check the active index because the active item may have been removed
+    const activeIndex = activeState ? getIndexOfOption(activeState) : -1;
+    let newActive = undefined;
+
+    // If the active item is still in the list, we don't need to do anything
+    if (activeIndex > 0) {
+      return;
+    }
+
+    // If we have selected an item, we should make that the active item
```
Comment (origami-z) [inline packages/core/src/list-box/ListBox.tsx:208]: This may not be covered in test

File: packages/core/src/menu/MenuGroup.css [hunk 4]
```diff
@@ -4,7 +4,7 @@
   font-size: var(--salt-text-label-fontSize);
   font-weight: var(--salt-text-label-fontWeight-strong);
   font-family: var(--salt-text-fontFamily);
-  line-height: var(--salt-text-lineHeight);
+  line-height: var(--salt-text-label-lineHeight);
```
Comment (origami-z) [inline packages/core/src/menu/MenuGroup.css:7]: This may need a changeset call out?

File: packages/core/src/list-box/ListBox.tsx [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+        newActive = getOptionAtIndex(Math.min(count, currentIndex + 10));
+        break;
+      case "Enter":
+      case " ":
+        if (
+          Boolean(activeState?.disabled) ||
+          (typeaheadString.current.trim().length > 0 && event.key === " ")
+        ) {
+          event.preventDefault();
+          return;
```
Comment (origami-z) [inline packages/core/src/list-box/ListBox.tsx:162]: Test: disabled option can be navigated but not selected

File: packages/core/src/list-box/ListBox.tsx (hunk 1 above)
Comment (joshwooding) [reply packages/core/src/list-box/ListBox.tsx:32]: It feels like a design difference, Lists by default have borders. We could always have bordered={false} but then we have to change our API standard

File: site/docs/components/list-box/accessibility.mdx [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+  $ref: ./#/title
+layout: DetailComponent
+sidebar:
+  label: Accessibility
+  exclude: true
+data:
+  $ref: ./#/data
+---
+
+## Keyboard interactions
```
Comment (origami-z) [inline site/docs/components/list-box/accessibility.mdx:12]: Does page up / down matter here?

File: site/docs/components/list-box/examples.mdx [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+The list box will scroll when the number of options exceeds the height of the list box.
+
+</LivePreview>
+
+<LivePreview componentName="list-box" exampleName="ComplexOptions">
+
+## Complex options
+
+The list box will scroll when the number of options exceeds the height of the list box.
```
Comment (origami-z) [inline site/docs/components/list-box/examples.mdx:65]: Need different text here?

File: site/docs/components/list-box/examples.mdx [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+Use the `disabled` prop on the `Option` component to disable an option. Disabled options are not selectable and are visually distinct from selectable options.
+
+</LivePreview>
+
+<LivePreview componentName="list-box" exampleName="Bordered">
+
+## Bordered
+
+Use the `borderless` prop to add a border to the list box.
```
Comment (origami-z) [inline site/docs/components/list-box/examples.mdx:49]: `borderless` Needs update

File: packages/core/stories/list-box/list-box.stories.css [hunk 9]
```diff
@@ -0,0 +1,4 @@
+.saltListBox {
```
Comment (origami-z) [inline packages/core/stories/list-box/list-box.stories.css:1]: Is it worth to be explicit with a different name? may cause debug confusion

File: site/src/examples/list-box/Scrolling.tsx [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+import { ReactElement } from "react";
+import { largestCities } from "./exampleData";
+
+export const Scrolling = (): ReactElement => {
+  return (
+    <ListBox
+      style={{
+        maxHeight:
+          "calc((var(--salt-size-base) + var(--salt-spacing-100)) * 8)",
+        width: "30%",
```
Comment (origami-z) [inline site/src/examples/list-box/Scrolling.tsx:11]: for site examples, is it better to use multiplier for size token to cater for different viewport?

File: site/src/examples/list-box/Scrolling.tsx [hunk 11]
```diff
@@ -0,0 +1,19 @@
+import { Option, ListBox } from "@salt-ds/core";
+import { ReactElement } from "react";
+import { largestCities } from "./exampleData";
```
Comment (origami-z) [inline site/src/examples/list-box/Scrolling.tsx:3]: With example data in another folder, could add an inline comment to github source?

**Review by origami-z — COMMENTED** [review]
Looking good after last change. Needing changesets before merge, around new list, option, option group CSS change, fixed Option disable

File: site/src/examples/list-box/Scrolling.tsx (hunk 10 above)
Comment (joshwooding) [reply site/src/examples/list-box/Scrolling.tsx:11]: cc @yunjungyeh

File: packages/core/src/__tests__/__e2e__/combo-box/ComboBox.cy.tsx [hunk 12]
```diff
@@ -255,20 +256,20 @@ describe("Given a ComboBox", () => {
...
     // should try to go down 10, but only 9 items in list
     cy.realPress(["PageDown"]);
-    cy.findAllByRole("option").eq(-1).should("be.activeDescendant");
-
-    // should not wrap
-    cy.realPress(["ArrowDown"]);
-    cy.findAllByRole("option").eq(-1).should("be.activeDescendant");
+    cy.findAllByRole("option").eq(14).should("be.activeDescendant");
 
     // should try to go up 10, but only 9 items in list
```
Comment (origami-z) [inline packages/core/src/__tests__/__e2e__/combo-box/ComboBox.cy.tsx:261]: Comment no longer apply I guess?

File: packages/core/src/combo-box/useComboBox.tsx [hunk 13]
```diff
@@ -70,7 +70,7 @@ export function useComboBox<Item>(props: UseComboBoxProps<Item>) {
     setSelectedState(newSelected);
     const newValue = getOptionsMatching((option) =>
       newSelected.includes(option.value)
-    ).map((option) => listControl.valueToString(option.value));
+    ).map((option) => listControl.valueToString(option.value.value));
```
Comment (origami-z) [inline packages/core/src/combo-box/useComboBox.tsx:73]: Don't know the intent for the change, `.value.value` means one of the variable naming is not great?

Comment (joshwooding) [reply packages/core/src/combo-box/useComboBox.tsx:73]: It's a breaking change to change it now

File: packages/core/src/list-box/ListBox.tsx [hunk 14]
```diff
@@ -0,0 +1,269 @@
...
+        onMouseDown={handleMouseDown}
+        ref={handleRef}
+        {...rest}
+      >
+        {children}
+      </div>
+    </ListControlContext.Provider>
+  );
+}) as <Item = string>(
+  props: ListBoxProps<Item> & { ref?: Ref<HTMLButtonElement> }
```
Comment (heswell) [inline packages/core/src/list-box/ListBox.tsx:268]: HTMLDivElement not HTMLButtonElement ?

File: packages/core/src/list-box/ListBox.tsx [hunk 15]
```diff
@@ -0,0 +1,269 @@
...
+  KeyboardEvent,
+  ReactNode,
+  Ref,
+  useRef,
+} from "react";
+import { useWindow } from "@salt-ds/window";
+import { useComponentCssInjection } from "@salt-ds/styles";
+import { clsx } from "clsx";
+import { makePrefixer, useForkRef } from "../utils";
+import { defaultValueToString } from "../list-control/ListControlState";
```
Comment (Fercas123) [inline packages/core/src/list-box/ListBox.tsx:15]:
```suggestion
```

File: packages/core/src/list-box/ListBox.tsx [hunk 16]
```diff
@@ -0,0 +1,269 @@
...
+import { ListControlContext } from "../list-control/ListControlContext";
+import {
+  useListControl,
+  ListControlProps,
+} from "../list-control/ListControlState";
```
Comment (Fercas123) [inline packages/core/src/list-box/ListBox.tsx:20]:
```suggestion
defaultValueToString,
} from "../list-control/ListControlState";
```

File: packages/core/src/option/OptionGroup.css [hunk 17]
```diff
@@ -6,11 +6,29 @@
...
+
+.saltOptionGroup {
+  border-top: var(--salt-size-border) var(--salt-separable-borderStyle) var(--salt-separable-tertiary-borderColor);
+}
+
+.saltOptionGroup:first-of-type {
+  border-top: 0;
+}
+
+/* Not implemented */
```
Comment (Fercas123) [inline packages/core/src/option/OptionGroup.css:23]: will this be implemented later or should it be cleaned?

**Review by Fercas123 — APPROVED** [review]

**Review by yunjungyeh — APPROVED** [review]

**Review by Fercas123 — APPROVED** [review]

**Review by yunjungyeh — APPROVED** [review]

**Pull request merged**